### DISCUSSION TITLE
How do I use my own car model?

### DISCUSSION THREAD

**Discussion by n1ckfg** [microsoft/AirSim#3300] [Support Q&A]
Looking to change the specs of the default car model--is there anything addressing this in the documentation?

**Comment by mitchellspryn** [comment]
If you're just looking to change the physics and are OK with keeping the same visual model, the simplest thing would be to edit the parameters of the included uasset file for the SUV (in AirSim Content/VehicleAdv/SUV/)

If you have a completely custom mesh, then you can set it up like any other wheeled vehicle, but instead parent it to CarPawn. And then change the CarPawn constructor to point to the correct assets.

**Comment by n1ckfg** [comment]
Thanks! Yup, just looking to change the physics, don't care about the appearance of the geo for now. Have you got anything there that would affect the throttle command? (Is there a drivetrain model?)

**Comment by mitchellspryn** [comment]
For throttle, there are a few things you can try. 

First, open the blueprint SuvCarPawn, and navigate to the VehicleMovement component:

![image](https://user-images.githubusercontent.com/20287427/33153705-3ca49f60-cf98-11e7-9b8f-07cac1934e14.png)

There are a few relevant parameters you can play with:

- Max RPM: this controls the top speed of the car. 
- Torque Curve: this controls the acceleration. To up it, add a few more points and slide it up. 

There are drivetrain / transmission models, but those usually affect handling (i.e. turning) more than acceleration. 

Do keep in mind that there will need to be some manual tuning; I have found that copy-pasting actual data from real-world tests done on car components (e.g. torque curves from engines) usually doesn't give very good results.

**Comment by n1ckfg** [comment]
Awesome, I'll play with this!

**Comment by benjamin-lange** [comment]
Would you please provide a more detailed tutorial on how add a custom mesh vehicle to AirSim? Thank you!

**Comment by chioveri** [comment]
Hi, I would also be happy to have more infos on how to use a custom mesh car/vehicle. Thank you

**Comment by benjamin-lange** [comment]
@chioveri 

May be this works for you, I will try it out as soon as I find the time:
https://github.com/Microsoft/AirSim/blob/master/docs/settings.md
The documentation deals with all sorts of parameters that can be
set in the "/Documents/Airsim/settings.json file". Have a look at the 
"pawn paths" section.

**Comment by fernandodamasio** [comment]
Hi guys,

I´m also having difficulties to change the mesh of the car. Just the mesh. Anyone knows how to do it? I tried different approaches like changing SpawnCar Path to a new blueprint and also trying to change only the Skeletal Mesh to a new Mesh.

If someone has already done this, could you exemplify how to do it?

![image](https://user-images.githubusercontent.com/21197744/47241943-c743e880-d3c3-11e8-9d1b-e101d08a0216.png)

![image](https://user-images.githubusercontent.com/21197744/47241832-746a3100-d3c3-11e8-9154-31a029e6788d.png)

Best regards,
Fernando Damasio

**Comment by benjamin-lange** [comment]
@fernandodamasio
I did not look into your code problem.
But here is a skeleton settings.json for the custom model I use:
`
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  "PawnPaths": {
	"DefaultCar": {"PawnBP":"Class'/Game/TruckTrailer/Truck/Blueprints/Truck.Truck_C'"}
	} 
`
You have to adjust the path in "DefaultCar" to your setup.
Please refer to the [documentation](https://github.com/Microsoft/AirSim/blob/master/docs/settings.md) for details. Look under "Other Settings" paragraph "PawnPaths".

**Comment by fernandodamasio** [comment]
> @fernandodamasio
> I did not look into your code problem.
> But here is a skeleton settings.json for the custom model I use:
> `"SettingsVersion": 1.2, "SimMode": "Car", "PawnPaths": { "DefaultCar": {"PawnBP":"Class'/Game/TruckTrailer/Truck/Blueprints/Truck.Truck_C'"} } `
> You have to adjust the path in "DefaultCar" to your setup.
> Please refer to the [documentation](https://github.com/Microsoft/AirSim/blob/master/docs/settings.md) for details. Look under "Other Settings" paragraph "PawnPaths".

Hi @benjamin-lange 

I tried to change only the mesh inside the cars' blueprint. It didn´t work. So, I guess I need another car to change it? I mean, I can't change just the mesh / appearance of the Airsim car.

Appreciate your help.

Best regards,
Fernando Damasio

**Comment by benjamin-lange** [comment]
@fernandodamasio , your are welcome.

> I tried to change only the mesh inside the cars' blueprint. It didn´t work.

I recommend leaving the default blueprint alone.
Just to clarify what you tried to do, you tried to change the mesh as it is shown in the following screenshot?

![image](https://user-images.githubusercontent.com/37362568/47415177-1cd91600-d773-11e8-84ed-c3b88e27869b.png)

If so then you might break the animations associated with the mesh which could cause more problems.
That is if you want to use the default vehicle in the future you have to fix these problems. 

> So, I guess I need another car to change it?

Yes, I recommend that. At least copy the original Blueprint located in project_name\Plugins\AirSim\Content\VehicleAdv\SUV and rename it before you experiment on it. ;-)
After you made your changes try out the pawn path method as stated above.

For your experiments with custom vehicle Blueprints in Airsim have a look into this [free asset ](https://www.unrealengine.com/marketplace/carsim-vehicle-dynamics) from the marktplace  or simply use the car model from the Blueprint vehicle template. To get the vehicle template create a new project using the Blueprint vehicle template and then extract the Blueprint from it. Then use the pawn paths in the settings to setup the vehicle Blueprint for Airsim. I think there is a way to import the vehicle template content into your current project without creating a new project as well, but I do not know the details for this procedure.

**Comment by fernandodamasio** [comment]
Hi @benjamin-lange 

Thank you very much for your quick response and also congrats for the Airsim project!

Regards,

**Comment by karanparikh91** [comment]
Hello!
Thanks,  @benjamin-lange , @mitchellspryn and everyone else who has contributed to solving this issue
I am attempting to change the default car in my AirSim project

Here's what I have done so far:

1. Imported my new fbx asset for the car. It has the same skeletal hierarchy and nomenclature as the SUV pawn.
2. Created a physics asset for it. This is set up in a similar way to the SUV blueprint which is the airsim default. The wheels are set to kinematic bodies and the chassis has the default physics setting instead of kinematic. I'm using single hull convex collisions on all bodies for my mesh 
3. Created a BP for my new skeletal mesh for the car. The BP I created has the CarPawn as its parent
4. Changed the Airsim settings.json file to include the pawn path to my new blueprint

When I hit play, I do see my new car show up instead of the default old one. However, the car doesn't seem to respond to any input when I hit the arrow keys. Am I missing a step or something else?

Thanks!

**Reply by GuihVicentini** [reply]
Hi there! 

Did you get this to work? 

I am facing the same issue. The animation even runs when I use the arrow keys, but the car doesn't move.

Thanks!

**Comment by benjamin-lange** [comment]
@karanparikh91, have you tried using the "WASD" keys to move the vehicle?

**Comment by karanparikh91** [comment]
@benjamin-lange  thanks for your reply.

Turns out the car was not exported properly from Blender. 
My colleague found this blog post useful in fixing the problem
blog post: https://forums.unrealengine.com/development-discussion/content-creation/66676-help-blender-vehicle-to-ue4

I've got things working correctly now.

**Comment by thejeshk** [comment]
Hi @benjamin-lange 
![my_own_vehicle_skeleton](https://user-images.githubusercontent.com/33362523/75112373-80af6f80-5643-11ea-8e93-ceab4dcaae5a.PNG)


I have my own own project and my own vehicle model with the following skeleton [see image]. I have created a child blueprint class from AIRSIM carpawn and selected a mesh of my own vehicle. What else I should do in addition to that to use my vehicle ?  


If I try to play. Iam getting the following error. [image] 
![load_error](https://user-images.githubusercontent.com/33362523/75112386-9b81e400-5643-11ea-988b-05fffc3fd6f5.PNG) 

![my_own_vehicle_skeleton](https://user-images.githubusercontent.com/33362523/75112388-a3418880-5643-11ea-98ce-428563530c3f.PNG)


My json settings are follows. 

{
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  "PawnPaths": {
    "DefaultCar": { "PawnBP": "Class'/Game/Mower/Blueprints/Mower_AIRSIM.Mower_AIRSIM'" }
  }
}

**Comment by thejeshk** [comment]
Hello @benjamin-lange, @karanparikh91 @mitchellspryn @fernandodamasio > 


I am having difficulties adding custom vehicle into my own project. 


Here's what I have done so far: Followed same steps as @karanparikh91 

**I have created my own project** 

1. Imported my new fbx asset for the car. **It has the same skeletal hierarchy but different bone names.** 
[Note: Is it important to have the same name bone names?] 
2. Has physics asset. Wheels are set to kinematic bodies. I'm using single hull convex collisions on all bodies for my mesh
3. Created a BP for my new skeletal mesh for the car. The BP I created has the CarPawn as its parent
4. Changed the Airsim settings.json file to include the pawn path to my new blueprint

When I hit play, I do see my new car flying from somewhere and dropping down (even though the player start is on the floor) But it doesn't seem to respond to any input when I hit the arrow keys. Am I missing a step or something else?

Thanks!
TK

**Reply by GuihVicentini** [reply]
Hi there!

Did you get this to work?

I am facing the same issue. The animation even runs when I use the arrow keys, but the car doesn't move.

Thanks!

**Reply by ssaigarimella** [reply]
Any update on this? Getting the same error.

**Reply by GuihVicentini** [reply]
Hi there!
I don’t remember it 100% but I made it work.
What I did was to import my car model in a clean UE project and make it work there first, then export it directly to the other project.

if I recall correctly, the wheel size was to big or too small. Even though the rendering was correct, the values were not, so the car was not moving.

There are some nice videos on YouTube how to import a car to UE from blender. My recommendation would be to check some of them, make the car fully work in an new UE project and then import it to your AirSim project

**Comment by VoltXd** [comment]
Hello, does anyone knows how to modify the little blue vehicule's (_\AirSim\Content\VehicleAdv\Vehicle_) size ? 
I tried to do it in _Vehicule_SkelMesh_ but the new configuration doesn't want to save.
In _VehicleAdvPawn_ i can change the scale but then the vehicule starts to act weird (high acceleration or steering, antenna flying at light speed, vehicule levitating), I assumed it was due to _VehiclePhysicsAsset_'s configuration but I don't know how this works.

Was I in the good way or there's an other mean to resize the car ?

**Reply by ssaigarimella** [reply]
Any update on fixing this, thanks
